### PR TITLE
custom allreduce + torch.compile

### PR DESCRIPTION
**Pull request by SageMoore**
This Pr changes pynccl all reduce to be out of place and removes support for torch distributed's all reduce.

### CODE REVIEW

File: vllm/distributed/parallel_state.py [hunk 1]
```diff
@@ -384,31 +365,23 @@ def all_reduce(self, input_: torch.Tensor) -> torch.Tensor:
...
+        #         out = ca_comm.custom_all_reduce(input_)
+        #         assert out is not None
+        #         return out
         pynccl_comm = self.pynccl_comm
-        if (pynccl_comm is not None and not pynccl_comm.disabled):
-            pynccl_comm.all_reduce(input_)
-        else:
-            torch.distributed.all_reduce(input_, group=self.device_group)
+        assert pynccl_comm is not None
+        with pynccl_comm.change_state(enable=True,
```
Comment (youkaichao) [inline vllm/distributed/parallel_state.py:380]: we can change `pynccl` to be always enabled.

**Comment by youkaichao** [comment]
should work now `vllm serve meta-llama/Llama-3.1-8B-Instruct -O3 -tp 2` :

```text
input_.shape=torch.Size([2048, 4096]), using pynccl allreduce
(VllmWorkerProcess pid=3908615) input_.shape=torch.Size([2048, 4096]), using pynccl allreduce


(VllmWorkerProcess pid=3908615) input_.shape=torch.Size([256, 4096]), using custom allreduce
input_.shape=torch.Size([248, 4096]), using custom allreduce
...
(VllmWorkerProcess pid=3908615) input_.shape=torch.Size([8, 4096]), using custom allreduce
input_.shape=torch.Size([2, 4096]), using custom allreduce
(VllmWorkerProcess pid=3908615) input_.shape=torch.Size([4, 4096]), using custom allreduce
input_.shape=torch.Size([1, 4096]), using custom allreduce
(VllmWorkerProcess pid=3908615) input_.shape=torch.Size([2, 4096]), using custom allreduce
INFO 11-25 16:46:25 custom_all_reduce.py:224] Registering 2275 cuda graph addresses
(VllmWorkerProcess pid=3908615) input_.shape=torch.Size([1, 4096]), using custom allreduce
(VllmWorkerProcess pid=3908615) INFO 11-25 16:46:26 custom_all_reduce.py:224] Registering 2275 cuda graph addresses
```

for profiling size `[2048, 4096]`, it is using pynccl.

for decode size `[256, 4096]`, it is using custom allreduce.

**Comment by youkaichao** [comment]
@SageMoore thanks for your pioneering investigation!

**Review by youkaichao — APPROVED** [review]

**Review by tlrmchlsmth — APPROVED** [review]
This looks good to me!

**Comment by SageMoore** [comment]
> @SageMoore thanks for your pioneering investigation!

Thanks for the help getting this over the line!

**Pull request merged**